### PR TITLE
[WIP] Prefect to Kafka Connector

### PR DESCRIPTION
**Pull request by leonkozlowski**
*WORK IN PROGRESS*
In b3d1a87:
Introduced the `python-kafka` module for eventual `.send(...)`

In e99b18c - fd01586:
-Created introductory `Flow` from `Task` subclass Objects
-Tore down the `main.py` as this is rather breaking change.
-Fixes to unit tests

**NOTE**
-There is a long `schema.json` in here for reference

*EDIT*
Moved docstrings to strict Google format [example](https://github.com/betfund/betfund-lines/pull/4#discussion_r402717036)
changed private instance variables to public
improved docstrings
increased verbosity
removed unused imports (which would have been used in future)
removed logging

### CODE REVIEW

File: lines/client/__init__.py [hunk 1]
```diff
@@ -1,6 +1,6 @@
 """BetFund client.client namespace."""
-from .rundown import RundownClient
+from .rundown import Rundown
 from .config import RundownSportId
 
 
-__all__ = ["RundownClient", "RundownSportId"]
+__all__ = ["Rundown", "RundownSportId"]
```
Comment (mitchbregs) [inline lines/client/__init__.py:6]: Why not -
```
__all__ = [Rundown, RundownSportId]
```

File: lines/client/config.py [hunk 2]
```diff
@@ -2,7 +2,16 @@
 from enum import Enum
 
 
-class RundownSportId(Enum):
+class ExtendedEnum(Enum):
+    """ExtendedEnum delegation.."""
```
Comment (mitchbregs) [inline lines/client/config.py]: Explain what this is doing.

File: lines/client/rundown.py [hunk 3]
```diff
@@ -1,57 +1,89 @@
 """Client for TheRundown API."""
```
Comment (mitchbregs) [inline lines/client/rundown.py]: Might be worth having this in a separate repo.  One repo for a client wrapper, the other a workflow service.

File: lines/client/rundown.py [hunk 4]
```diff
@@ -1,57 +1,89 @@
...
+            )
 
-        url_extras = "sports/{}/events/{}".format(
-            str(sport_id), self._today.strftime("%Y-%m-%d")
+        now = datetime.datetime.now()
+        request_url = "{}/sports/{}/events/{}".format(
+            self.base_url, str(sport), now.strftime("%Y-%m-%d")
+        )
+
+        _LOGGER.info(
```
Comment (mitchbregs) [inline lines/client/rundown.py]: I'm going to create a separate logging utility, but ok for now.

File: lines/flow/fetch.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+from lines.client.rundown import Rundown
+from lines.producer.sender import KafkaProducer
+from lines.transformer.transform import RundownTransformer
+
+
+def build(sport: int) -> Flow:
+    """Build flow via imperative API."""
+    rdc = Rundown()
+    rtc = RundownTransformer()
+    kpdr = KafkaProducer()
```
Comment (mitchbregs) [inline lines/flow/fetch.py]: Preferential: verbosity.

```
rundown_client = Rundown()
...
```

File: lines/flow/fetch.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    with Flow("betfund-lines") as flow:
+        sport = Parameter("sport")
+
```
Comment (mitchbregs) [inline lines/flow/fetch.py:30]: For all of these `set_dependencies`, can we add a comment of what is happening?

File: lines/flow/fetch.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        )
+
+        flow.set_dependencies(
+            task=kpdr, mapped=True, keyword_tasks=dict(record=(rtc))
+        )
+
+    return flow
+
+
+def execute(flow: Flow, sport: int) -> State:
```
Comment (mitchbregs) [inline lines/flow/fetch.py:53]: Can we add executor handling?

File: lines/main.py [hunk 8]
```diff
@@ -1,29 +1,11 @@
...
-from lines.client import RundownClient
-from lines.snapshot import TemporalToSnapshot
-from lines.temporal import BetFundTemporalClient
+from lines.client import Rundown  # noqa: F403, F401
+from lines.transformer import RundownTransformer  # noqa: F403, F401
 
 
-def run(sport_id: int):  # pragma: no cover
+def run(sport_id: int):
     DEBUG = False  # DO NOT CHECK IN AS TRUE
```
Comment (mitchbregs) [inline lines/main.py]: What's going on here?

File: lines/producer/sender.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+from prefect import Task
+
+from lines.libs.errors import MalformedPayloadError
+
+
+class EventProducer(Task):
+    """Event Producer Object."""
+
+    def __init__(self, *args, **kwargs):
+        self._server = os.getenv("KAFKA_BOOTSTRAP_SERVER")  # temporary
```
Comment (mitchbregs) [inline lines/producer/sender.py]: Docstring:

```
"""
This handles producing new events to the Kafka pipeline.  Typically, these events will be xyz.

Attributes
----------
_server (str): Server address for Kafka.
_topic (str): Kafka topic for producing events.
"""
```

Comment (mitchbregs) [reply lines/producer/sender.py]: Why use `_xyz`? These seem very important.

File: lines/producer/sender.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+        self._topic = os.getenv("KAFKA_TOPIC")
+        super().__init__(*args, **kwargs)
+
+    def run(self, record: dict):
+        """Implements `KafkaProducer.send(...)`."""
```
Comment (mitchbregs) [inline lines/producer/sender.py]: Explain what is happening here. Include comments for each action.

File: lines/producer/sender.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        producer = self._build_producer()
+
+        response = producer.send(
+            topic=self._topic, value=record
+        )
+
+        return response
+
+    def _build_producer(self):
+        """Builder of KafkaProducer."""
```
Comment (mitchbregs) [inline lines/producer/sender.py]: Why not 
```
def build(self):  # we already know that this this is the KafkaProducer
    """Explain what is happening"""
    ...
```

File: lines/producer/sender.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+class EventProducer(Task):
+    """Event Producer Object."""
+
+    def __init__(self, *args, **kwargs):
+        self._server = os.getenv("KAFKA_BOOTSTRAP_SERVER")  # temporary
+        self._topic = os.getenv("KAFKA_TOPIC")
+        super().__init__(*args, **kwargs)
+
+    def run(self, record: dict):
```
Comment (mitchbregs) [inline lines/producer/sender.py:29]: Might be better named `send`?

File: lines/transformer/transform.py [hunk 13]
```diff
@@ -0,0 +1,78 @@
+"""
```
Comment (mitchbregs) [inline lines/transformer/transform.py:1]: Absolutely love this.

File: lines/transformer/transform.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+            "eventId": "9a35b8986a76ea364be331cb453ec",  # string
+            "sportId": 2  # integer
+        },
+        "data": {
+            "timestamp": "2020-02-01T17:22:12Z",
+            "eventDate": "2020-02-02T23:30:00Z",
+            ...
+        }
+    }
+"""
```
Comment (mitchbregs) [inline lines/transformer/transform.py:24]: Should we use `jsonschema` to validate fields/payload? Link [here](https://python-jsonschema.readthedocs.io/en/stable/).

File: lines/transformer/transform.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        """Constructor for RundownTransformer."""
+        self.log_level = "INFO"
+        super().__init__()
+
+    def run(self, record: LinesEventFacade) -> Union[dict, None]:
+        """
+        Generate temporal record for data store.
+
+        Args:
+            record: (LinesEventFacade) - input to generate temporal record for
```
Comment (mitchbregs) [inline lines/transformer/transform.py]: Let's stick to a consistent docstring pattern. If you use Google, use it as [such](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html).

File: tests/unit/test_client.py [hunk 16]
```diff
@@ -1,53 +1,43 @@
...
-from lines.client.rundown import RundownClient
+from lines.client.rundown import Rundown
+from lines.libs.errors import InvalidSportIdError
+
 from tests.helpers import MockRequestsResponse
 
 
-class TestRundownClient(TestCase):
-    """Unit tests for RundownClient."""
+class TestRundown(TestCase):
```
Comment (mitchbregs) [inline tests/unit/test_client.py:14]: Love it, love it, love it.

File: lines/client/__init__.py (hunk 1 above)
Comment (mitchbregs) [reply lines/client/__init__.py:6]: Same for many of the `init`'s below.

File: lines/client/rundown.py [hunk 17]
```diff
@@ -1,57 +1,89 @@
...
 from lines.client.config import RundownSportId
-from lines.libs import LinesResponseFacade
+from lines.libs.errors import InvalidSportIdError
+from lines.libs.facades import LinesResponseFacade
+
+logging.basicConfig(level="INFO")
+_LOGGER = logging.getLogger(__name__)
+
+
+class Rundown(Task):
```
Comment (mitchbregs) [inline lines/client/rundown.py:13]: Love the class `Task`.  I've been hating `@task` recently, 😄.

File: lines/client/rundown.py [hunk 18]
```diff
@@ -1,57 +1,89 @@
...
 
+from prefect import Task
+
 from lines.client.config import RundownSportId
-from lines.libs import LinesResponseFacade
+from lines.libs.errors import InvalidSportIdError
+from lines.libs.facades import LinesResponseFacade
+
+logging.basicConfig(level="INFO")
+_LOGGER = logging.getLogger(__name__)
```
Comment (mitchbregs) [inline lines/client/rundown.py]: For now, let's separate to a `logger.py`. I will replace with `betfund_logging` package.

File: lines/client/rundown.py [hunk 19]
```diff
@@ -1,57 +1,89 @@
...
+        sport: (int) - sport_id value of `RundownSportId`
+    Returns:
+        lines: (LinesResponseFacade) - facade of response object `.json()`
+    """
 
-    def __repr__(self):
-        """Representation for RundownClient."""
-        return "<RundownClient: {}>".format(str(self._today))
+    def __init__(self, *args, **kwargs):
+        """Constructor for Rundown"""
```
Comment (mitchbregs) [inline lines/client/rundown.py:30]:
```
Attributes
----------
xyz (str): Does such and such.
```

Not sure if `----` is used in Google docstrings.  I've been liking `numpy` recently.

File: lines/client/rundown.py [hunk 20]
```diff
@@ -1,57 +1,89 @@
...
+        if sport not in RundownSportId._value2member_map_:
+            raise InvalidSportIdError(
+                "Valid `sport` values: {}".format(
+                    RundownSportId.list()
+                )
+            )
 
-        url_extras = "sports/{}/events/{}".format(
-            str(sport_id), self._today.strftime("%Y-%m-%d")
+        now = datetime.datetime.now()
```
Comment (mitchbregs) [inline lines/client/rundown.py:57]: Can use `datetime.date` as well.

Let's add comments.  Ex:

```
# Calling datetime to get date for latest sporting lines
now = datetime.datetime.now()

# Build the request URL for Rundown Events API
request_url = ...
```

File: lines/client/rundown.py [hunk 21]
```diff
@@ -1,57 +1,89 @@
...
         params = {
-            "include": ["all_periods", "scores"],
+            "include": [
+                "all_periods",
+                "scores"
+            ],
             "offset": "0"
         }
 
         headers = {
```
Comment (mitchbregs) [inline lines/client/rundown.py:70]: Can be held in the `__init__`, your call.

File: lines/client/rundown.py [hunk 22]
```diff
@@ -1,57 +1,89 @@
...
             url=request_url, headers=headers, params=params
         )
         response.raise_for_status()
 
-        return LinesResponseFacade(response.json())
+        lines = LinesResponseFacade(
+            response.json()
+        )
+
+        return lines.events
```
Comment (mitchbregs) [inline lines/client/rundown.py:84]: Why the change here?

File: lines/flow/fetch.py [hunk 23]
```diff
@@ -0,0 +1,63 @@
+"""Flow builder."""
+import os  # noqa: F403, F401
+from datetime import timedelta  # noqa: F403, F401
+
+from prefect import Flow, Parameter
+from prefect.engine.executors import DaskExecutor  # noqa: F403, F401
+from prefect.engine.state import State
+from prefect.schedules import IntervalSchedule  # noqa: F403, F401
+from prefect.utilities.debug import raise_on_exception  # noqa: F403, F401
```
Comment (mitchbregs) [inline lines/flow/fetch.py]: Is this ever being called?

File: lines/flow/fetch.py [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+
+from lines.client.rundown import Rundown
+from lines.producer.sender import KafkaProducer
+from lines.transformer.transform import RundownTransformer
```
Comment (mitchbregs) [inline lines/flow/fetch.py:7]: Really appreciate all the relative imports.

File: lines/transformer/transform.py [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+                "linePeriods": record.line_periods,
+            }
+        }
+
+        if not all([record.event_id, record.sport_id]):
+            raise MalformedKeyError(
+                "Invalid Key: {}".format(
+                    json.dumps(transformed.get("key"), indent=4)
+                )
+            )
```
Comment (mitchbregs) [inline lines/transformer/transform.py:78]: This is a pretty wild format.  Not sure what you are going for.

**Review by mitchbregs — CHANGES_REQUESTED** [review]
I'm sorry for the super nitpicky review.  I really want well organized, documented, verbose code.  It allows for others to pick up on your work that much easier.  Really solid overall, I know this works well because I trust your engineering. Great work.  Let me know if you want to discuss any of this.  A lot of these comments are preferential.

File: lines/client/__init__.py (hunk 1 above)
Comment (leonkozlowski) [reply lines/client/__init__.py:6]: https://docs.python.org/3/tutorial/modules.html#importing-from-a-package

File: lines/producer/sender.py (hunk 11 above)
Comment (leonkozlowski) [reply lines/producer/sender.py]: This is exactly what is happening

File: lines/producer/sender.py (hunk 12 above)
Comment (leonkozlowski) [reply lines/producer/sender.py:29]: Subclasses of `Task` are executed by `run(...)`

File: lines/flow/fetch.py (hunk 24 above)
Comment (leonkozlowski) [reply lines/flow/fetch.py:7]: these are absolute imports

File: lines/client/config.py (hunk 2 above)
Comment (leonkozlowski) [reply lines/client/config.py]:
```python
    @classmethod
    def list(cls):
        """List ExtendedEnum for Enum."""
        return list(map(lambda c: c.value, cls))
```

The above provides an interface for `Exception` to provide all allowed values for a `RundownSportId`

File: lines/producer/sender.py (hunk 9 above)
Comment (leonkozlowski) [reply lines/producer/sender.py]: the leading underscore is used as these are considered private instance variables

File: lines/transformer/transform.py (hunk 25 above)
Comment (leonkozlowski) [reply lines/transformer/transform.py:78]: What is wild about it?
Providing a pretty printed version of key in exception for readability

File: lines/client/rundown.py (hunk 22 above)
Comment (leonkozlowski) [reply lines/client/rundown.py:84]: `.map()` requires an iterable input

File: lines/transformer/transform.py (hunk 15 above)
Comment (leonkozlowski) [reply lines/transformer/transform.py]: from google docstring documentation:
```
def function_with_types_in_docstring(param1, param2):
    """Example function with types documented in the docstring.

    `PEP 484`_ type annotations are supported. If attribute, parameter, and
    return types are annotated according to `PEP 484`_, they do not need to be
    included in the docstring:

    Args:
        param1 (int): The first parameter.
        param2 (str): The second parameter.

    Returns:
        bool: The return value. True for success, False otherwise.

    .. _PEP 484:
        https://www.python.org/dev/peps/pep-0484/

    """
```

File: lines/client/rundown.py (hunk 21 above)
Comment (leonkozlowski) [reply lines/client/rundown.py:70]: Would prefer to not include function/endpoint specific variables in `__init__`

File: lines/client/config.py (hunk 2 above)
Comment (mitchbregs) [reply lines/client/config.py]: Write that in the docstring!

File: lines/client/rundown.py (hunk 22 above)
Comment (mitchbregs) [reply lines/client/rundown.py:84]: Seems like a separate API client repository is worth it.  I wouldn't sacrifice your API client happenings for Prefect.

File: lines/flow/fetch.py (hunk 24 above)
Comment (mitchbregs) [reply lines/flow/fetch.py:7]: Yup, confused terminology lol.

File: lines/client/rundown.py (hunk 22 above)
Comment (leonkozlowski) [reply lines/client/rundown.py:84]: Going to scrap this then

File: lines/producer/sender.py (hunk 11 above)
Comment (mitchbregs) [reply lines/producer/sender.py]: Why not call is `build()`.

File: lines/producer/sender.py (hunk 12 above)
Comment (mitchbregs) [reply lines/producer/sender.py:19]: Yup, totally my bad.

File: lines/transformer/transform.py (hunk 15 above)
Comment (mitchbregs) [reply lines/transformer/transform.py]: Yes.. hence:

```
record (LinesEventFacade): Input to generate temporal record.
```

File: lines/producer/sender.py (hunk 11 above)
Comment (leonkozlowski) [reply lines/producer/sender.py]: build from where?

Comment (leonkozlowski) [reply lines/producer/sender.py]: this is a private method, only called within a public method

File: lines/producer/sender.py (hunk 9 above)
Comment (mitchbregs) [reply lines/producer/sender.py]: Should they be considered private?

File: lines/producer/sender.py (hunk 11 above)
Comment (mitchbregs) [reply lines/producer/sender.py]: Gotcha

**Review by mitchbregs — APPROVED** [review]
This looks awesome 🎉
